### PR TITLE
refactor: Concentrate knowledge of directory locations

### PR DESCRIPTION
**Pull request by jskeet**

### CODE REVIEW

**Comment by jskeet** [comment]
Fixes b/383482997

File: tools/Google.Cloud.Tools.ReleaseManager/GenerateApisCommand.cs [hunk 1]
```diff
@@ -181,13 +185,13 @@ void ValidateNamespaces()
 
     private void GenerateGapicApi(ApiMetadata api)
     {
-        var layout = DirectoryLayout.ForApi(api.Id, generatorOutputDirectory, generatorInputDirectory);
-        string productionDirectory = Path.Combine(layout.SourceDirectory, api.Id);
+        var apiLayout = _rootLayout.CreateApiLayout(api);
+        string productionDirectory = Path.Combine(apiLayout.SourceDirectory, api.Id);
```
Comment (amanda-tarafa) [inline tools/Google.Cloud.Tools.ReleaseManager/GenerateApisCommand.cs:189]: I think this is the same as a built package directory I flagged for moving into ApiLayout a few comments up. If it's here again, maybe it does make sense to move it to ApiLayout.

File: tools/Google.Cloud.Tools.ReleaseManager/GenerateApisCommand.cs [hunk 2]
```diff
@@ -251,21 +255,21 @@ private void GenerateGapicApi(ApiMetadata api)
...
             }
         }
     }
 
     private void GenerateProtoApi(ApiMetadata api)
     {
-        var layout = DirectoryLayout.ForApi(api.Id, generatorOutputDirectory, generatorInputDirectory);
-        string productionDirectory = Path.Combine(layout.SourceDirectory, api.Id);
+        var apiLayout = _rootLayout.CreateApiLayout(api);
+        string productionDirectory = Path.Combine(apiLayout.SourceDirectory, api.Id);
```
Comment (amanda-tarafa) [inline tools/Google.Cloud.Tools.ReleaseManager/GenerateApisCommand.cs:269]: And here again, doesn't ApiLayout already knows about the API ID.

**Review by amanda-tarafa — APPROVED** [review]
This is great!

All of my comments are about more refactoring on the same line, i.e. moving more paths into RootLayout or ApiLayout. I'm on the fence about some of them as it would require they layout types to know about special folders and files, and even default values for some of those. I think that's fine, and I think having the layout types and only the layout types know everything about folders and files is fine. But I'm also fine with the not knowing about many specific things.

Any changes you want to make, happy to have those in other PRs.

File: tools/Google.Cloud.Tools.GenerateDocfxSources/Program.cs [hunk 3]
```diff
@@ -285,8 +287,9 @@ private static List<string> GetClientClasses(ApiMetadata api)
             {
                 return new List<string>();
             }
-            var layout = DirectoryLayout.ForApi(api.Id);
-            var packageSource = Path.Combine(layout.SourceDirectory, api.Id);
+            var rootLayout = RootLayout.ForCurrentDirectory();
+            var apiLayout = rootLayout.CreateApiLayout(api);
+            var packageSource = Path.Combine(apiLayout.SourceDirectory, api.Id);
```
[the remark this answers is not in the document]
Comment (jskeet) [reply tools/Google.Cloud.Tools.GenerateDocfxSources/Program.cs:292]: Yes. (Or at least, it does at build time.) I think we've got this in multiple places, in fact - certainly in generation. Will do as a follow-up. I'll try finding every occurrence of Path.Combine to see whether there are other similar patterns, too.

File: tools/Google.Cloud.Tools.GenerateSnippetMarkdown/Program.cs [hunk 4]
```diff
@@ -140,17 +140,18 @@ private static int MainImpl(string[] args)
                 }
             }
 
-            DirectoryLayout layout = api == "help" ? DirectoryLayout.ForHelpSnippets() : DirectoryLayout.ForApi(api);
+            var rootLayout = RootLayout.ForCurrentDirectory();
+            var apiLayout = api == "help" ? ApiLayout.ForHelpSnippets(rootLayout) : rootLayout.CreateApiLayout(api);
```
[the remark this answers is not in the document]
Comment (jskeet) [reply tools/Google.Cloud.Tools.GenerateSnippetMarkdown/Program.cs:144]: I strongly suspect this is the only occurrence of this - we may want to move the logic into this file, if we can.

File: tools/Google.Cloud.Tools.ReleaseManager/ContainerCommands/CleanCommand.cs [hunk 5]
```diff
@@ -80,12 +79,12 @@ void CleanApiFiles(ApiMetadata api)
                     DeleteGeneratedSource("");
                     DeleteGeneratedSource(".Snippets");
                     DeleteGeneratedSource(".GeneratedSnippets");
-                    var generatedSnippetsDirectory = Path.Combine(layout.SourceDirectory, $"{api.Id}.GeneratedSnippets");
+                    var generatedSnippetsDirectory = Path.Combine(apiLayout.SourceDirectory, $"{api.Id}.GeneratedSnippets");
```
[the remark this answers is not in the document]
Comment (jskeet) [reply tools/Google.Cloud.Tools.ReleaseManager/ContainerCommands/CleanCommand.cs:82]: Hmm... will have a think. (Again, I'll see how many places we use Path.Combine for this.)

File: tools/Google.Cloud.Tools.ReleaseManager/ContainerCommands/CleanCommand.cs [hunk 6]
```diff
@@ -80,12 +79,12 @@ void CleanApiFiles(ApiMetadata api)
...
                     if (Directory.Exists(generatedSnippetsDirectory))
                     {
                         DeleteAll(Directory.EnumerateFiles(generatedSnippetsDirectory, "*.json"));
                     }
-                    File.Delete(Path.Combine(layout.SourceDirectory, "gapic_metadata.json"));
+                    File.Delete(Path.Combine(apiLayout.SourceDirectory, "gapic_metadata.json"));
```
[the remark this answers is not in the document]
Comment (jskeet) [reply tools/Google.Cloud.Tools.ReleaseManager/ContainerCommands/CleanCommand.cs:87]: Ooh, a method like that would be nice, yes.

File: tools/Google.Cloud.Tools.ReleaseManager/ContainerCommands/CleanCommand.cs [hunk 7]
```diff
@@ -95,11 +94,11 @@ void CleanApiFiles(ApiMetadata api)
...
-            PruneEmptyDirectories(layout.SourceDirectory);
+            PruneEmptyDirectories(apiLayout.SourceDirectory);
 
             void DeleteGeneratedSource(string directorySuffix)
             {
-                var directory = Path.Combine(layout.SourceDirectory, $"{api.Id}{directorySuffix}");
+                var directory = Path.Combine(apiLayout.SourceDirectory, $"{api.Id}{directorySuffix}");
                 if (Directory.Exists(directory))
                 {
                     DeleteAll(Directory.EnumerateFiles(directory, "*.g.cs", SearchOption.AllDirectories));
```
[the remark this answers is not in the document]
Comment (jskeet) [reply tools/Google.Cloud.Tools.ReleaseManager/ContainerCommands/CleanCommand.cs:104]: Unsure at the moment. Bear in mind that quite a bit of that is gapic-specific.

File: tools/Google.Cloud.Tools.ReleaseManager/ContainerCommands/GenerateCommand.cs [hunk 8]
```diff
@@ -42,17 +42,14 @@ public int Execute(Dictionary<string, string> options)
...
-        // Note: we expect the container to already have environment variables for
-        // protoc, protobuf tools root, the gRPC generator, and the GAPIC generator.
-
         // For unconfigured generation, we use a really minimal generator-input directory in the container itself.
         // At the time of writing, this is only used for the common resource config (which hadn't changed between March 2021
         // and December 2024, so is clearly pretty stable).
-        Environment.SetEnvironmentVariable(GenerateApisCommand.GeneratorInputDirectoryEnvironmentVariable, generatorInput ?? "/app/generator-input");
-        Environment.SetEnvironmentVariable(GenerateApisCommand.GeneratorOutputDirectoryEnvironmentVariable, output);
-        Environment.SetEnvironmentVariable(GenerateApisCommand.GoogleApisDirectoryEnvironmentVariable, apiRoot);
+        var rootLayout = RootLayout.ForGeneration(generatorInput ?? "/app/generator-input", output, apiRoot);
```
[the remark this answers is not in the document]
Comment (jskeet) [reply tools/Google.Cloud.Tools.ReleaseManager/ContainerCommands/GenerateCommand.cs:48]: I don't think so in this case - it's a really special case of wanting to use /app, which is very specific to the container.

File: tools/Google.Cloud.Tools.ReleaseManager/GenerateApisCommand.cs (hunk 1 above)
Comment (jskeet) [reply tools/Google.Cloud.Tools.ReleaseManager/GenerateApisCommand.cs:189]: Yes, exactly what I was thinking :)

**Comment by jskeet** [comment]
Merging, but will do a follow-up on Monday.

**Pull request merged**